jfr-streaming/src/main/java/io/opentelemetry/contrib/jfr/metrics/internal/memory/G1HeapSummaryHandler.java

```diff
@@ -0,0 +1,117 @@
+  private static final String METRIC_DESCRIPTION_MEMORY = "Heap utilization";
+  private static final String EVENT_NAME = "jdk.G1HeapSummary";
+  private static final String BEFORE = "Before GC";
+  private static final String AFTER = "After GC";
+  private static final String GC_ID = "gcId";
+  private static final String WHEN = "when";
+  private static final Attributes ATTR_MEMORY_EDEN_USED = Attributes.of(ATTR_USAGE, EDEN_USED);
+  private static final Attributes ATTR_MEMORY_EDEN_SIZE = Attributes.of(ATTR_USAGE, EDEN_SIZE);
+  private static final Attributes ATTR_MEMORY_EDEN_SIZE_DELTA =
+      Attributes.of(ATTR_USAGE, EDEN_SIZE_DELTA);
+  private static final Attributes ATTR_MEMORY_SURVIVOR_SIZE =
+      Attributes.of(ATTR_USAGE, SURVIVOR_SIZE);
+  private static final Attributes ATTR_MEMORY_REGIONS = Attributes.of(ATTR_USAGE, REGION_COUNT);
+
+  private final Map<Long, RecordedEvent> awaitingPairs = new HashMap<>();
+
+  private DoubleHistogram memoryHistogram;
+
+  public G1HeapSummaryHandler() {
+    initializeMeter(defaultMeter());
+  }
+
+  @Override
+  public void initializeMeter(Meter meter) {
+    memoryHistogram =
+        meter
+            .histogramBuilder(METRIC_NAME_MEMORY)
+            .setDescription(METRIC_DESCRIPTION_MEMORY)
+            .setUnit(BYTES)
+            .build();
+  }
+
+  @Override
+  public String getEventName() {
+    return EVENT_NAME;
+  }
+
+  @Override
+  public void accept(RecordedEvent ev) {
+    String when = null;
+    if (ev.hasField(WHEN)) {
+      when = ev.getString(WHEN);
+    }
+    if (when != null) {
+      if (!(when.equals(BEFORE) || when.equals(AFTER))) {
+        return;
+      }
+    }
+
+    if (!ev.hasField(GC_ID)) {
+      return;
+    }
+    long gcId = ev.getLong(GC_ID);
+
+    var pair = awaitingPairs.get(gcId);
+    if (pair == null) {
+      awaitingPairs.put(gcId, ev);
+    } else {
+      awaitingPairs.remove(gcId);
+      if (when != null && when.equals(BEFORE)) {
+        recordValues(ev, pair);
+      } else { //  i.e. when.equals(AFTER)
+        recordValues(pair, ev);
+      }
+    }
+  }
+
+  private void recordValues(RecordedEvent before, RecordedEvent after) {
+    if (after.hasField("edenUsedSize")) {
+      memoryHistogram.record(after.getLong("edenUsedSize"), ATTR_MEMORY_EDEN_USED);
+      if (before.hasField("edenUsedSize")) {
+        memoryHistogram.record(
+            after.getLong("edenUsedSize") - before.getLong("edenUsedSize"),
+            ATTR_MEMORY_EDEN_SIZE_DELTA);
+      }
+    }
+    if (after.hasField("edenTotalSize")) {
+      memoryHistogram.record(after.getLong("edenTotalSize"), ATTR_MEMORY_EDEN_SIZE);
+    }
+    if (after.hasField("survivorUsedSize")) {
+      memoryHistogram.record(after.getLong("survivorUsedSize"), ATTR_MEMORY_SURVIVOR_SIZE);
+    }
+    if (after.hasField("numberOfRegions")) {
+      memoryHistogram.record(after.getLong("numberOfRegions"), ATTR_MEMORY_REGIONS);
+    }
```

Comment:
What's the interpretation of the delta of eden size?

What's the thought behind using a histogram here with these dimensions? Each of the dimensions seems like a distinct stream - it doesn't appear there is anything useful you can do by aggregating across the dimensions.

Reply:
That's a fair point. Tbh, these will probably be refactored to UpDownCounters as part of the next version.